### PR TITLE
Log javaOptions/fork interactions

### PR DESCRIPTION
**Pull request by pdalpra**
This fixes both #2041 and #2087.
- When forking, log the javaOptions that are used
- When javaOptions are defined but fork := false, warn that javaOptions
  will be ignored

### CODE REVIEW

**Comment by typesafe-tools** [comment]
Can one of the admins verify this patch?

**Comment by dwijnand** [comment]
1) This requires notes.
2) You're breaking bincompat changing those public methods
3) It's interesting to note the different approach here compared to https://github.com/Duhemm/sbt/commit/fc6a3373058c6b7d7c7a492b05196ba2318b41d0, ie. logging as late as possible with what information percolates down (Martin's way), vs logging as early as possible at ever entry point (your way). It's a trade-off between information availability and making sure a log always happen. I'm interested in what other think.

**Comment by pdalpra** [comment]
1) Indeed, I waited for the PR creation, since the PR number is often included in the notes
2) Argh, I'll need to see if I can get the javaOptions in another way to allGroupTestTasks (or somewhere else).
3) On the topic of logging the javaOptions that are used, @Duhemm 's solution is without hesitation a better solution, since it avoids code duplication. However, this can't work for 2041, since the issue is that we're not forking but javaOptions are still provided and ignored.

File: main/src/main/scala/sbt/Defaults.scala [hunk 1]
```diff
@@ -541,20 +541,24 @@ object Defaults extends BuildCommon {
     }
   }
 
-  def allTestGroupsTask(s: TaskStreams, frameworks: Map[TestFramework, Framework], loader: ClassLoader, groups: Seq[Tests.Group], config: Tests.Execution, cp: Classpath, javaHome: Option[File]): Task[Tests.Output] = {
-    allTestGroupsTask(s, frameworks, loader, groups, config, cp, javaHome, forkedParallelExecution = false)
+  def allTestGroupsTask(s: TaskStreams, frameworks: Map[TestFramework, Framework], loader: ClassLoader, groups: Seq[Tests.Group], config: Tests.Execution, cp: Classpath, javaHome: Option[File], javaOptions: Seq[String]): Task[Tests.Output] = {
```
Comment (jsuereth) [inline main/src/main/scala/sbt/Defaults.scala]: This isn't private, so we can't add an option to it without also created an overloaded method that has the original # of arguments (binary compatibility)

File: main/src/main/scala/sbt/Defaults.scala [hunk 2]
```diff
@@ -541,20 +541,24 @@ object Defaults extends BuildCommon {
...
+    allTestGroupsTask(s, frameworks, loader, groups, config, cp, javaHome, javaOptions, forkedParallelExecution = false)
   }
 
-  def allTestGroupsTask(s: TaskStreams, frameworks: Map[TestFramework, Framework], loader: ClassLoader, groups: Seq[Tests.Group], config: Tests.Execution, cp: Classpath, javaHome: Option[File], forkedParallelExecution: Boolean): Task[Tests.Output] = {
+  def allTestGroupsTask(s: TaskStreams, frameworks: Map[TestFramework, Framework], loader: ClassLoader, groups: Seq[Tests.Group], config: Tests.Execution, cp: Classpath, javaHome: Option[File], javaOptions: Seq[String], forkedParallelExecution: Boolean): Task[Tests.Output] = {
```
Comment (jsuereth) [inline main/src/main/scala/sbt/Defaults.scala]: Similar here.

Comment (jsuereth) [inline main/src/main/scala/sbt/Defaults.scala]: Also, can we just use `opts.runJVMOptions` to check if they are empty, rather than pass down into this function?

File: main/src/main/scala/sbt/Defaults.scala [hunk 3]
```diff
@@ -739,7 +743,17 @@ object Defaults extends BuildCommon {
...
     val si = scalaInstance.value
-    if (fork.value) new ForkRun(forkOptions.value) else new Run(si, trapExit.value, tmp)
+    val s = streams.value
+    val options = javaOptions.value
+    if (fork.value) {
+      s.log.debug(s"javaOptions: $options")
+      new ForkRun(forkOptions.value)
+    } else {
+      if (options.nonEmpty) {
+        s.log.warn("javaOptions will be ignored, fork is set to false")
```
Comment (jsuereth) [inline main/src/main/scala/sbt/Defaults.scala]: Thanks so much for adding this, will help a lot.

**Comment by jsuereth** [comment]
@dwijnand You win for reviewing while I had this open and waiting, thereby misisng your comments completely.  Agree with your comments.

File: main/src/main/scala/sbt/Defaults.scala (hunk 2 above)
Comment (pdalpra) [inline main/src/main/scala/sbt/Defaults.scala]: See below, this is what I do when the `runPolicy` is `SubProcess`, as I have access to the `ForkOptions`.
Unfortunately, I cannot do that when the Strategy is `InProcess`...

**Comment by pdalpra** [comment]
So, should I remove the `javaOptions` part and leave it to @Duhemm, or should I implement his solution by logging in `ForkJava`/`LocalJava` ?

**Comment by pdalpra** [comment]
Added notes.

**Comment by dwijnand** [comment]
@pdalpra Do we log the same (or "sufficiently" similar) information and in all the same cases by moving it to `ForkJava`/`LocalJava`? If so then I think as late as possible is better.

**Comment by pdalpra** [comment]
@dwijnand By moving the logging to `ForkJava`/`LocalJava`, we can only fix #2087 : we're forking and we log the options that are used.
#2041 needs to be done upfront, since we're not forking.

@jsuereth @dwijnand I just pushed a fix for the bincompat issue, by addind an overload for `allGroupTestTasks`.

**Comment by dwijnand** [comment]
Doesn't `LocalJava` deal with the non-forked version?

**Comment by pdalpra** [comment]
`LocalJava` seems to deal only with compilation :/

**Comment by pdalpra** [comment]
More precisely, `LocalJava` goal (as stated by its documentation : `Helper methods for trying to run the java toolchain out of our own classloaders.`) is to run the Javadoc tool or `javac` in process.
After all, running a main class or tests in process boils down to calling a method.

**Comment by pdalpra** [comment]
BTW, what @Duhemm has done on his fork does not affects the `run` or `test` tasks, as what he has done logs the JVM options used for `javac`/`javadoc`.

File: main/src/main/scala/sbt/Defaults.scala (hunk 3 above)
Comment (jsuereth) [inline main/src/main/scala/sbt/Defaults.scala]: Also, thinking about this, it's not actually telling WHICH CONFIGURATION has javaOptions and fork set to false.  You may actually want to grab the magic "what's my current configuration" dependency so you can report which one is the problem, otherwise it'll just be popping out and giving the user no idea how to solve it.  GIve me a abit and I'll look up the syntax.

Comment (jsuereth) [inline main/src/main/scala/sbt/Defaults.scala]: Ok, it looks like the best you can do is depend on the `sbt.Keys.configuration` and grab the currently active configuration.  We also need you to grab the scoped "task", ideally, so you can tell users where the bad setting is.

**Comment by jsuereth** [comment]
BC changes look good, was just commenting on what the error messages look like in practice (From a quick test).

Ideally I'd like to know, if I set `fork in (Compile, run) := false` and `javaOptions in Compile := Seq(...)` I'd like to know WHICH `runner` is the problematic one.

**Comment by pdalpra** [comment]
Indeed, as it is, It's nothing more than a hint that something is wrong for any not-completely-trivial build.
Will fix ASAP.

**Comment by pdalpra** [comment]
I have two small questions: 
- if we are to print the configuration, this mean that we need to add yet another parameter to `allGroupTestTasks`, which is not so hard in itself. However, as we'll need to provide a default value for that parameter, which configuration should be used as the default value ?
- Is there a simple way to get the name of the Scope ? I found `Scope.display`, should I use it ?

File: main/src/main/scala/sbt/Defaults.scala (hunk 3 above)
Comment (pdalpra) [inline main/src/main/scala/sbt/Defaults.scala]: I'm still missing some info to get the scoped task, but that's where I am right now :

```
[warn] */*:javaOptions will be ignored, */*:fork is set to false
```

**Comment by jsuereth** [comment]
So, as for allGroupTestTasks... hmm... Maybe let's defer that for now.  I think there may be an alternative way to get the error message on that setting.

As for Scope.display, that should be fine.   I personally get confused over the different show options EVERY TIME  I need to show something, so you should try it out and check the error messages.   if it's wrong, we'll find you the right one :)

**Comment by jsuereth** [comment]
Also, thanks @pdalpra for the dedication on this one.  Very glad to see you going the extra mile on error messages :)

**Comment by pdalpra** [comment]
As shown above, I managed to get something that looks quite good, except for the scopes...
Is `javaOptions.scopedKey` the right way to get the key with its scope ?

You're welcome :) 
After all, if we're going to add log messages, let's do it right, no ? ;)

**Comment by jsuereth** [comment]
no, I dont think `javaOptions.scopedKey` is doing the right thing...

**Comment by pdalpra** [comment]
It seems that what is done in `SettingsGraph` (used by `inspect tree`) could get me to the info I want, but it looks like it's overkill just to get scoped `run` and `javaOptions` :/

**Comment by jsuereth** [comment]
@pdalpra I'll have to look. The issue is NOT that you want the scopedKey of the current key, it's that you want to find the scope of the current Setting you're a part of.

SO, for ANY key, you can do:  `key.scopedKey`.  This is because, internally, sbt uses ScopedKeys for its setting system.  A scoped key is just an `AttributeKey` and a `Scope`.  `TaskKey` and `SettingKey` and friends are the user-facing DSL for these.  They allow you to manipulate them and also keep the high level API for things.

If you just call `javaOptions.scopedKey` you'll get a `ScopeKey` with the default scope.   If you're inspecting a setting _AFTER_ setting resolution, you'll see the real scope for the key (we do translation/resolution during a build-compile step).

SO yes, `SettingGraph` is looking at the Finalized `ScopedKey` (our internal API).  What you see in your setting is _before_ this step, so the scopedKey would be wrong.  

WHat we do have in sbt, are a few "magic" keys that will be injected by the setting system.  For example, `taskStreams` is a setting which gets injected based on the scope of the key the setting is being applied to.   We expose _ONE_ of these _magic_ keys called "configuration" which you can use to grab the current configuration you're being applied to.   See:  https://github.com/sbt/sbt/blob/0.13/main/src/main/scala/sbt/Load.scala#L142-L168

TL;DR;  Try to depend on `resolvedScope` key and that should give you the right scope.

**Comment by pdalpra** [comment]
Thank you for your detailed explanation, I'll try it ASAP.

**Comment by pdalpra** [comment]
Got it : 

```
[warn] {file:/Users/pdalpra/test/}test/compile:run::javaOptions will be ignored, {file:/Users/pdalpra/test/}test/compile:run::fork is set to false
```

I only logged scopes on the warning, not on the debug that display the javaOptions.

Can't get that level of details to `allGroupTestTasks` however.

**Comment by pdalpra** [comment]
Amended my commit so that the project axis in not shown in the warning message :

```
[warn] compile:run::javaOptions will be ignored, compile:run::fork is set to false
```

**Comment by jsuereth** [comment]
@pdalpra nice!

**Comment by dwijnand** [comment]
Nice indeed!

**Comment by pdalpra** [comment]
BTW, thank you for your patience with me on this @jsuereth !
